fennel/test_lib/mock_client.py

```diff
@@ -188,18 +209,40 @@ def dataset_lookup_impl(
         right_df = data[cls_name]
         right_df[FENNEL_LOOKUP] = True
         right_df[FENNEL_TIMESTAMP] = right_df[timestamp_field]
+        cols_to_replace = []
         for col in keys:
             if col in right_df and keys[col].dtype != right_df[col].dtype:
                 keys[col] = keys[col].astype(right_df[col].dtype)
+            # Changing dtype of Struct to str for making it hashable
```

Comment:
The lookup function has now become extremely complex and over 150 lines long. Might be a good time to break it up, and maybe move it into a module where you can reuse code, the if and else branch can clearly use some DRY

Reply:
For now I have some code to a different function. Will pick a task to modularize the code.